File: ps_1_to_100/p10.cpp
// brute force would not work
// use sieve of eratosthenes

#include <iostream>
#include <cmath>
#include <vector>

using namespace std;
bool is_prime(int n){
    
    if (n == 1){
        return false;
    }
    else if (n % 2 == 0 and n != 2){
        return  false;
    }
    for(int i=3; i <= sqrt(n); i+=2){
        if (n%i==0){
            return false;
        }
    }

    return true;
}

int main(){
    int num = 1;
    long sum_prime = 0; 
    int req_primes = 2000000;
    while ( num <= req_primes)
    {
        if (is_prime(num)) {
            // cout << num << " ";
            sum_prime += num;
            num++;
            if (num==req_primes){
                break;
            }
        }
        else
        {
            num++;
        }
    }
    cout << "\nThe sum of the primes below " << req_primes << " is: " << sum_prime << endl;
    return 1;
}

File: ps_1_to_100/python/p10.py
# print sum of all primes below 200,000,0
# have to use eratosthenes sieve for extracting 
# primes
# Implementation based on `python3` `set()` function
# is enormously faster. It uses hash-table for lookup
# which is highly efficient. 


import numpy as np 
from copy import deepcopy

def is_prime(n, prime_list):
    if n==2 or n==3:
        return True
    for i in prime_list:
        if i != 0:
            if (n % i ==0):
                return False
    return True


def main_program(n):
    multiples = set()
    for i in range(2, n+1):
        if i not in multiples:
            yield i
            multiples.update(range(i*i, n+1, i))
 
# print(list(eratosthenes2(100)))

print(sum(main_program(2000000)))